### PR TITLE
feat(avoidance): implement avoidance path safety check logic

### PR DESCRIPTION
**Pull request by satoshi-ota**
### BackGround

The previous avoidance module did not have a safety check mechanism, and the operator had no choice but to OR (over ride) and interrupt the avoidance maneuver as manualy if a vehicle in the adjacent or opposite lane approached after the module had switched output paths for avoidance. This problems has a significant impact on miles per disengagement, so we would like to implement safety check mechanism and yield behavior in the module.

As a first step, I implement safetye check mechanism in this PR. The latter (yield maneuver) will be implement in another PR.

### Logic

Add following new parameters for safety check mechanism.

```yaml
enable_safety_check: false

# For safety check
safety_check_backward_distance: 50.0  # [m]
safety_check_time_horizon: 10.0 # [s]
safety_check_idling_time: 1.5 # [s]
safety_check_accel_for_rss: 2.5  # [m/ss]
```
#### Step1

`safety_check_backward_distance` is the parameter related to the safety check area. The module checks a collision risk for all vehicle that is within shift side lane and between object _check_forward_distance ahead and safety_check_backward_distance behind.

**NOTE**: Even if a part of an object polygon overlaps the detection area, if the center of gravity of the object does not exist on the lane, the vehicle is excluded from the safety check target.

image

#### Step2

Judge the risk of collision based on ego future position and object prediction path. The module calculates Ego's future position in the time horizon (`safety_check_time_horizon`), and use object's prediction path as object future position.

image

After calculating the future position of Ego and object, the module calculates the lateral/longitudinal deviation of Ego and the object. The module also calculates the lateral/longitudinal margin necessary to determine that it is safe to execute avoidance maneuver, and if both the lateral and longitudinal distances are less than the margins, it determines that there is a risk of a collision at that time.

image

The value of the longitudinal margin is calculated based on Responsibility-Sensitive Safety theory ([RSS](https://newsroom.intel.com/articles/rss-explained-five-rules-autonomous-vehicle-safety/#gs.ljzofv)). The `safety_check_idling_time` represents $T_{idle}$, and `safety_check_accel_for_rss` represents $a_{max}$.

```math
D_{lon} = V_{ego}T_{idle} + \frac{1}{2}a_{max}T_{idle}^2 + \frac{(V_{ego} + a_{max}T_{idle})^2}{2a_{max}} - \frac{V_{obj}^2}{2a_{max}}
```

The lateral margin is changeable based on ego longitudinal velocity. If the vehicle is driving at a high speed, the lateral margin should be larger, and if the vehicle is driving at a low speed, the value of the lateral margin should be set to a smaller value. Thus, the lateral margin for each vehicle speed is set as a parameter, and the module determines the lateral margin from the current vehicle speed as shown in the following figure.

```yaml
target_velocity_matrix:
  col_size: 4
  matrix:
    [2.78, 5.56, 11.1, 13.9,   # target velocity [m/s]
     0.20, 0.90, 1.10, 1.20]   # margin [m]
```

image

## Related links

- https://github.com/tier4/autoware_launch/pull/637

### Evaluator Result

[[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/056ebb1d-3d8a-52ad-9a36-12400596f07d?project_id=prd_jt) This PR#1 (817/825)
[[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/431498f3-9e2e-5533-89a2-d2b00b83467d?project_id=prd_jt) This PR#2 (818/825)
[[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/981ab539-bd85-5a45-9fe2-781255a71554?project_id=prd_jt) Baseline (812/825)

### Test in local

Befor the tests, please use [this](https://github.com/tier4/autoware_launch/pull/637) autoware_launch branch and set following params in config to `true`.

```yaml
enable_safety_check: true
publish_debug_marker: true # optional
```

Since this PR only implements a function that returns the safety of the avoidance path as a boolean, it does not change the behavior of the vehicle (Even if the module judges the avoidance path is NOT safe, do nothing. The vehicle will go on avoidance maneuver.). On the other hand, we can see the status via debug marker and check whether the module has determined that the avoidance path is safe or not.

rviz_screenshot_2022_12_20-19_26_01

rviz_screenshot_2022_12_20-19_26_20

- debug marker topic: `/planning/scenario_planning/lane_driving/behavior_planning/behavior_path_planner/debug/avoidance`
- The vehicles that has collision risk are shown as BLUE CUBE in rviz (namespace `unsafe_objects`).
- The safety judge is shown as WHITE TEXT in rviz (namespace `ego_status`)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by TakaHoribe — COMMENTED** [review]
>NOTE: Even if a part of an object polygon overlaps the detection area, if the center of gravity of the object does not exist on the lane, the vehicle is excluded from the safety check target.

Would you let me know the reason of this approach.

**Comment by satoshi-ota** [comment]
> > NOTE: Even if a part of an object polygon overlaps the detection area, if the center of gravity of the object does not exist on the lane, the vehicle is excluded from the safety check target.
> 
> Would you let me know the reason of this approach.

@TakaHoribe Thanks for your comment.
The reason why we use CoG of object instead of polygon is that we do **NOT** want to extract vehicles in the ego's lane as safety check target as much as possible. With the current object recognition accuracy, even if a vehicle traveling behind ego,  its polygon sometimes have overlap with the adjacent lane, so if the target is filtered by the former method, the safety check mechanism may make wrong dicision as unsafe for vehicle that is in ego's lane. On the other hand, the module can estimate the object driving lane relatively accurately by using CoG.

File: planning/behavior_path_planner/src/utilities.cpp [hunk 1]
```diff
@@ -2538,4 +2538,115 @@ double calcLaneChangeBuffer(
   return num_lane_change * calcTotalLaneChangeDistanceWithBuffer(common_param) +
          length_to_intersection;
 }
+
+bool applyVelocitySmoothing(
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/utilities.cpp:2542]: It is not recommended to use the smoother here. Currently, smoother is used in a few nodes, but we do not want to add more from a code management point. Now we can't change the role of the smoother easily due to the dependencies, which is the drawback.
I assume the reason for using the smoother is to accurately predict the future velocity, but there is a limitation of the hierarchical architecture. We should solve this issue in the architecture design.

For now, in the avoidance module, the future ego velocity can be assumed as a constant, or increasing with a constant acceleration up to the velocity limit.

Comment (satoshi-ota) [reply planning/behavior_path_planner/src/utilities.cpp:2542]: OK. I'll remove smoother.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp [hunk 2]
```diff
@@ -496,4 +498,132 @@ void generateDrivableArea(
...
+    }
+
+    const auto expand_lane = lanelet::utils::getExpandedLanelet(lane, l_offset, r_offset);
+    target_lanelets.push_back(expand_lane);
+  }
+
+  return target_lanelets;
+}
+
+std::vector<size_t> filterObjectIndicesByLanelets(
```
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp:562]: As avoidance module, we need not only avoidance target object but also non avoidance target object for avoidance safety check. Therefore, this function is designed to return objects that do **NOT** satisfy the avoidance target conditions as a vector.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 3]
```diff
@@ -1939,6 +1988,370 @@ void AvoidanceModule::addReturnShiftLineFromEgo(
...
+
+  auto path_with_current_velocity = shifted_path.path;
+  path_with_current_velocity = util::resamplePathWithSpline(path_with_current_velocity, 0.5);
+
+  const size_t ego_idx = findEgoIndex(path_with_current_velocity.points);
+  util::clipPathLength(path_with_current_velocity, ego_idx, forward_check_distance, 0.0);
+
+  constexpr double MIN_EGO_VEL_IN_PREDICTION = 1.38;  // 5km/h
+  for (auto & p : path_with_current_velocity.points) {
+    p.point.longitudinal_velocity_mps = std::max(getEgoSpeed(), MIN_EGO_VEL_IN_PREDICTION);
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2013]: @TakaHoribe Predicts future position with the assumption that the vehicle will keep its current speed.

File: planning/behavior_path_planner/src/utilities.cpp (hunk 1 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/utilities.cpp]: Removed smoother.

**Comment by satoshi-ota** [comment]
Add evaluator result in description.

> [[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/056ebb1d-3d8a-52ad-9a36-12400596f07d?project_id=prd_jt) This PR (817/825)
[[TIER IV INTERNAL]](https://evaluation.tier4.jp/evaluation/reports/981ab539-bd85-5a45-9fe2-781255a71554?project_id=prd_jt) Baseline (812/825)

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp [hunk 4]
```diff
@@ -496,4 +498,132 @@ void generateDrivableArea(
...
+  for (const auto & llt : target_lanelets) {
+    if (boost::geometry::within(object_centroid, llt.polygon2d().basicPolygon())) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+lanelet::ConstLanelets getTargetLanelets(
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp:525]: This function is not used anywhere. Is this ok?

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 5]
```diff
@@ -1939,6 +1988,370 @@ void AvoidanceModule::addReturnShiftLineFromEgo(
...
+{
+  if (path.points.empty()) {
+    return true;
+  }
+
+  const auto path_with_time = [&path]() {
+    std::vector<std::pair<PathPointWithLaneId, double>> ret{};
+
+    float travel_time = 0.0;
+    ret.push_back(std::make_pair(path.points.front(), travel_time));
```
Comment (TakaHoribe) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:2034]: Minor: this is the case `emplace_back` is preferred.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp [hunk 6]
```diff
@@ -220,14 +225,23 @@ void AvoidanceModule::fillAvoidanceTargetObjects(
   // detection area filter
   // when expanding lanelets, right_offset must be minus.
   // This is because y axis is positive on the left.
-  const auto expanded_lanelets = lanelet::utils::getExpandedLanelets(
-    data.current_lanelets, parameters_->detection_area_left_expand_dist,
+  const auto expanded_lanelets = getTargetLanelets(
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp:228]: @TakaHoribe `getTargetLanelets` is used here.

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp (hunk 5 above)
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_module.cpp]: fix in https://github.com/autowarefoundation/autoware.universe/pull/2534/commits/9b086265914106761f7e8b2ce8d4063c828915cf

File: planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp]: Remove this function from avoidance_util.hpp, and add new function in utilities.hpp.

[the remark this answers is not in the document]
Comment (satoshi-ota) [reply planning/behavior_path_planner/src/scene_module/avoidance/avoidance_utils.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/2534/commits/d9e023d14ad756caa68bc786ec13bbae29cf95a9

**Review by TakaHoribe — APPROVED** [review]

**Comment by satoshi-ota** [comment]
@TakaHoribe Thanks for all your reviews. I have wrote release not. I'll merge this.

**Comment by yukkysaito** [comment]
@satoshi-ota After this PR is merged, maybe Psim cannot work :pray: 
image

**Comment by satoshi-ota** [comment]
> @satoshi-ota @TakaHoribe After this PR is merged, maybe Psim cannot work pray image

@yukkysaito Thanks for your report. I'm sorry about that. I will be careful. :bow:

**Pull request merged**